### PR TITLE
Feature: Detect masterkey change

### LINKED ISSUE #142: "Outside" changes to masterkey and vault config are not considered
### Please agree to the following

- [X] I have searched [existing issues](https://github.com/cryptomator/cryptomator/issues?q=) for duplicates
- [X] I agree to follow this project's [Code of Conduct](https://github.com/cryptomator/cryptomator/blob/develop/.github/CODE_OF_CONDUCT.md)

### Summary

I created vault on local storage on my Iphone and just for testing, i choose to delete masterkey.cryptomator file but in iPhone Files App, i can access and decrypt files which contains my vault without the key file.

thanks for your support ! :)

### What software is involved?

- Operating System: IOS 15.1.1
- Cryptomator: 2.0.4
- use Cryptomator on File App of IOS

### Volume Type

FUSE

### Steps to Reproduce

1. 
Create a vault 

2. 
Delete the file key « masterkey.cryptomator » in the vault

3. 
decrypt with face id or password the vault (you can access it without the key file)

### Expected Behavior

an error because the file master key is not found

### Actual Behavior

i can access to files, the file "masterkey.cryptomator" is detect as not found only when i remove the vault and try to add it a new time on app

### Reproducibility

Always

### Relevant Log Output

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by phil1995**
Refreshes the cached vault during each manual unlock and thus fixes #142.

If there is currently no internet connection, the already cached vault is used.

It may now happen that the password stored in the keychain for the biometric unlock is no longer correct. In this case, a special error message is displayed to the user, the old password is removed from the keychain and the enable biometric unlock switch is already set to on for the next unlock.

At the same time, another customized error message was introduced if the user cancels the biometric unlock.

Note: The customized authentication errors are now possible in the FileProviderExtension, because we now use undocumented FileProviderErrorKeys (googling these error keys currently does not even lead to a single search result).

### CODE REVIEW

File: CryptomatorCommon/Sources/CryptomatorCommonCore/Manager/VaultDBCache.swift [hunk 1]
```diff
@@ -68,4 +75,95 @@ public class VaultDBCache: VaultCache {
...
+		}.recover { error -> Void in
+			if case CloudProviderError.itemNotFound = error {
+				try self.setVaultConfigData(nil, forVaultUID: vault.vaultUID, lastModifiedDate: nil)
+			}
+		}
+	}
+
+	private func masterkeyFileHasLegacyVaultFormat(_ masterkeyFile: MasterkeyFile) -> Bool {
+		return masterkeyFile.version < 8
+	}
```
Comment (tobihagemann) [inline CryptomatorCommon/Sources/CryptomatorCommonCore/Manager/VaultDBCache.swift:138]: This seems unused.

File: CryptomatorCommon/Sources/CryptomatorCommonCore/Manager/VaultDBManager.swift [hunk 2]
```diff
@@ -382,7 +411,7 @@ public class VaultDBManager: VaultManager {
 	}
 
 	func postProcessChangePassphrase(masterkeyFileData: Data, forVaultUID vaultUID: String, vaultConfigToken: Data?, newPassphrase: String) throws {
-		let cachedVault = CachedVault(vaultUID: vaultUID, masterkeyFileData: masterkeyFileData, vaultConfigToken: vaultConfigToken, lastUpToDateCheck: Date())
+		let cachedVault = CachedVault(vaultUID: vaultUID, masterkeyFileData: masterkeyFileData, vaultConfigToken: vaultConfigToken, lastUpToDateCheck: Date(), masterkeyFileLastModifiedDate: nil, vaultConfigLastModifiedDate: nil)
```
Comment (tobihagemann) [inline CryptomatorCommon/Sources/CryptomatorCommonCore/Manager/VaultDBManager.swift:414]: Should these dates really be set to `nil`? Wouldn't this cause the cache to be invalidated with the next unlock?

**Review by tobihagemann — CHANGES_REQUESTED** [review]

**Review by tobihagemann — CHANGES_REQUESTED** [review]
I did the following test with a vault "On My iPhone" (simulator), which I copied from my Mac:

1. Unlock vault and then lock vault (just to make sure that the cached vault is up-to-date)
2. Change password of that vault on my Mac
3. Delete `masterkey.cryptomator` on iOS and add the new `masterkey.cryptomator` from step 2 (basically, replace the masterkey file)
4. Unlock vault with new password

At first, the unlock is not successful because of "wrong password" (shaking screen). But if I press "Unlock" again (without changing anything, still using the new password), the unlock is successful.

It seems to me that the cached vault is not invalidated soon enough.

**Comment by phil1995** [comment]
> I did the following test with a vault "On My iPhone" (simulator), which I copied from my Mac:
> 
> 1. Unlock vault and then lock vault (just to make sure that the cached vault is up-to-date)
> 2. Change password of that vault on my Mac
> 3. Delete `masterkey.cryptomator` on iOS and add the new `masterkey.cryptomator` from step 2 (basically, replace the masterkey file)
> 4. Unlock vault with new password
> 
> At first, the unlock is not successful because of "wrong password" (shaking screen). But if I press "Unlock" again (without changing anything, still using the new password), the unlock is successful.
> 
> It seems to me that the cached vault is not invalidated soon enough.

Indeed the vault cache was refreshed / updated too late, because it is already accessed in the `FileProviderExtensionUI` (to derive the KEK) and the cache was not refreshed until the derived KEK was already passed via XPC.

Should be fixed now - the vault cache is now already updated in the `FileProviderExtensionUI` (in the `UnlockVaultViewModel`) so that the KEK is actually derived from the current version of the masterkey file.

**Review by tobihagemann — CHANGES_REQUESTED** [review]
Since errors during an unlock due to remote changes is now a thing, we should use the `LocalizedCloudProviderDecorator` so that the error messages are more meaningful. :smile:

I noticed that while testing a vault, where I removed the masterkey file.

**Review by tobihagemann — APPROVED** [review]

**Pull request merged**